javatests/arcs/android/storage/service/StorageServiceNgImplTest.kt

```diff
@@ -0,0 +1,176 @@
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.storage.service
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import arcs.android.crdt.ParcelableCrdtType
+import arcs.android.storage.service.testing.FakeMessageCallback
+import arcs.android.storage.service.testing.FakeStorageChannelCallback
+import arcs.android.storage.toParcelToByteArray
+import arcs.core.data.CountType
+import arcs.core.storage.FixedDriverFactory
+import arcs.core.storage.StorageKey
+import arcs.core.storage.StorageKeyParser
+import arcs.core.storage.StoreOptions
+import arcs.core.storage.UntypedProxyMessage
+import arcs.core.storage.keys.RamDiskStorageKey
+import arcs.core.storage.testutil.MockDriverProvider
+import arcs.core.storage.testutil.testWriteBackProvider
+import arcs.core.type.Type
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class StorageServiceNgImplTest {
+
+  private val driverFactory = FixedDriverFactory(MockDriverProvider())
+  private val onProxyMessageCallback: suspend (StorageKey, UntypedProxyMessage) -> Unit =
+    { _: StorageKey, _: UntypedProxyMessage -> }
+  private val messageCallback = FakeMessageCallback()
+  private val channelCallback = FakeStorageChannelCallback()
+
+  private val scope = TestCoroutineScope()
+  private lateinit var storageService: StorageServiceNgImpl
+
+  @Before
+  fun setUp() {
+    StorageKeyParser.addParser(RamDiskStorageKey)
+    storageService = StorageServiceNgImpl(
+      scope,
+      driverFactory,
+      ::testWriteBackProvider,
+      null,
+      onProxyMessageCallback
+    )
+  }
+
+  @After
+  fun tearDown() {
+    scope.cleanupTestCoroutines()
+  }
+
+  @Test
+  fun openStorageChannel_invokesOnCreateCallback_andCreatesChannel() = scope.runBlockingTest {
+    val encodedStoreOptions = createEncodedStoreOptions(
+      RamDiskStorageKey("myCount"),
+      CountType(),
+      ParcelableCrdtType.Count
+    )
+
+    storageService.openStorageChannel(encodedStoreOptions, channelCallback, messageCallback)
+
+    // `waitForOnCreate` ensures the `onCreate` callback was invoked.
+    val channel = channelCallback.waitForOnCreate()
+
+    assertThat(channel).isInstanceOf(StorageChannelImpl::class.java)
+  }
+
+  @Test
+  fun openingTwoChannels_forSameStore_resultsInTwoDistinctChannels() = scope.runBlockingTest {
+    val encodedStoreOptions = createEncodedStoreOptions(
+      RamDiskStorageKey("myCount"),
+      CountType(),
+      ParcelableCrdtType.Count
+    )
+
+    val channelCallback2 = FakeStorageChannelCallback()
+    storageService.openStorageChannel(encodedStoreOptions, channelCallback, messageCallback)
```

Comment:
Optional: This is more of a convention or personal preference than anything else, but usually I like to pair up my variables like this. So if I define `channelCallback2` I would also define `channelCallback1` to go with it. As it is now, `channelCallback` is a member variable, whereas `channelCallback2` is a local variable, and as a reader of this test method you kind of wonder why they are different.  Unless there's a lot of boilerplate to set them up, I usually prefer to define another local `channelCallback1` and just not use the member variable at all in the test.